lib/auth/user.go

```diff
@@ -209,11 +224,54 @@ func (s *Server) DeleteUser(ctx context.Context, user string) error {
 		},
 		UserMetadata: ClientUserMetadata(ctx),
 		ResourceMetadata: apievents.ResourceMetadata{
-			Name: user,
+			Name: username,
 		},
 	}); err != nil {
 		log.WithError(err).Warn("Failed to emit user delete event.")
 	}
 
 	return nil
 }
+
+// checkUserRoleConstraint checks if the request will result in having
+// no users with access to update roles.
+func (s *Server) checkUserRoleConstraint(ctx context.Context, user types.User, request string) error {
+	rolesWithUpdateRolesRule, err := s.getRolesWithUpdateRolesRule(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	allUsers, err := s.Identity.GetUsers(false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```

Comment:
IMO let's avoid fetching all roles and users in the cluster until we at least check that the user is losing a role

Reply:
See updated code